lgc/builder/BuilderImpl.cpp

```diff
@@ -74,6 +74,165 @@ Value *BuilderImplBase::CreateDotProduct(Value *const vector1, Value *const vect
+
+  // The result type is given by accumulator, which must be greater than or equal to that of the components of Vector 1
+  Type *resultTy = accumulator->getType();
+  const bool hasAccumulator = !(isa<ConstantInt>(accumulator) && cast<ConstantInt>(accumulator)->isNullValue());
+  const unsigned resBitWidth = resultTy->getScalarSizeInBits();
+  assert(resBitWidth == 4 || resBitWidth == 8 || resBitWidth == 16 || resBitWidth == 32 || resBitWidth == 64);
+
+  // Check if there is a native intrinsic that can do the entire operation (dot product and saturating accumulate) in a
+  // single instruction. They must meet the two conditions:
+  // 1. The required native intrinsic is supported by the specified hardware
+  // 2. The factor inputs must be <2 x i16> or <N x i8> (N <= 4) or <N x i4> (N <= 8)
+  const auto &supportIntegerDotFlag = getPipelineState()->getTargetInfo().getGpuProperty().supportIntegerDotFlag;
+  const bool isSupportCompBitwidth = (supportIntegerDotFlag.compBitwidth16 && compBitWidth == 16) ||
+                                     (supportIntegerDotFlag.compBitwidth8 && compBitWidth == 8) ||
+                                     (supportIntegerDotFlag.compBitwidth4 && compBitWidth == 4);
+  const bool isSupportSignedness =
+      isMixedSigned ? supportIntegerDotFlag.diffSignedness : supportIntegerDotFlag.sameSignedness;
+  const bool isDot2 = (compCount == 2 && compBitWidth == 16);
+  const bool isDot4 = (compCount <= 4 && compBitWidth == 8);
+  const bool isDot8 = (compCount <= 8 && compBitWidth == 4);
+  const bool hasNativeIntrinsic = isSupportCompBitwidth && isSupportSignedness && (isDot2 || isDot4 || isDot8);
+
+  // Create a lamda to clamp the input as expected
+  auto doClamp = [&](Value *input, Type *expectedTy) {
+    // Saturating
+    const unsigned resultWidth = expectedTy->getScalarSizeInBits();
+    const unsigned inputWidth = input->getType()->getScalarSizeInBits();
+    // Compute the clamp range based on the
+    unsigned long long unsignedMax = (2ULL << (resultWidth - 1)) - 1;
+    long long signedMax = unsignedMax >> 1;
+    long long signedMin = -1LL - signedMax;
+
+    Value *minimum = nullptr, *maximum = nullptr;
+    Value *isUnderflow = nullptr, *isOverflow = nullptr;
+    if (isSigned) {
+      minimum = ConstantInt::getSigned(input->getType(), signedMin);
+      maximum = ConstantInt::getSigned(input->getType(), signedMax);
+      isUnderflow = CreateICmpSLT(input, minimum);
+      isOverflow = CreateICmpSGT(input, maximum);
+    } else {
+      minimum = getIntN(inputWidth, 0);
+      maximum = getIntN(inputWidth, unsignedMax);
+      isUnderflow = CreateICmpULT(input, minimum);
+      isOverflow = CreateICmpUGT(input, maximum);
+    }
+    input = CreateSelect(isUnderflow, minimum, input);
+    input = CreateSelect(isOverflow, maximum, input);
+    return CreateTrunc(input, expectedTy);
+  };
```

Comment:
It would be cleaner to not use a lambda here anymore, since it is only used once.

Also, it now occurred to me that we should be using the `smin`/`smax`/`umin`/`umax` intrinsics here. You can use `CreateBinaryIntrinsic` for that.

Reply:
Thank you for the good suggestions. Will fix.

Reply:
Fixed.